gdx-jnigen/src/main/java/com/badlogic/gdx/jnigen/build/toolchains/BaseToolchain.java

```diff
@@ -0,0 +1,214 @@
+package com.badlogic.gdx.jnigen.build.toolchains;
+
+import com.badlogic.gdx.jnigen.BuildConfig;
+import com.badlogic.gdx.jnigen.BuildTarget;
+import com.badlogic.gdx.jnigen.FileDescriptor;
+import com.badlogic.gdx.jnigen.build.RuntimeEnv;
+import com.badlogic.gdx.jnigen.build.ToolchainExecutor;
+import com.badlogic.gdx.utils.TargetType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.*;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+
+public abstract class BaseToolchain {
+
+    private static final Logger logger = LoggerFactory.getLogger(BaseToolchain.class);
+
+    protected RuntimeEnv ENV = new RuntimeEnv();
+    protected BuildConfig config;
+    protected BuildTarget target;
+    protected List<File> sourceCollectedCFiles;
+    protected List<File> sourceCollectedCppFiles;
+
+    protected List<File> buildCFiles;
+    protected List<File> buildCppFiles;
+
+    protected File buildDirectory;
+    protected File libsDirectory;
+    protected File targetLibFile;
+
+    public void configure (BuildTarget target, BuildConfig config) {
+        this.target = target;
+        this.config = config;
+        this.buildDirectory = config.buildDir.child(target.getTargetFolder()).file();
+        this.libsDirectory = config.libsDir.child(target.getTargetFolder()).file();
+
+        this.targetLibFile = target.getTargetBinaryFile(config);
+
+    }
+
+    public abstract void checkForTools ();
+
+    public void build () {
+        checkForTools();
+
+        collectCFiles();
+        collectCPPFiles();
+        cleanBuildDirectory();
+        createBuildDirectory();
+
+        compile();
+    }
+
+    public abstract void compile ();
+
+    private void collectCFiles () {
+        File rootDir = config.jniDir.file();
+        if (!rootDir.exists()) {
+            throw new RuntimeException("Jni directory does not exist at path " + config.jniDir.file().getAbsolutePath());
```

Comment:
Yeah, thinking more about it, the cIncludes are really difficult to work with now, since you still can't do releative path like `../`, but now you are in the build directory, where you can't commit files into.

Reply:
The jni dir should always exist at this point, because to run the toolchain you should have generated with jnigen, which copies headers and makes cpp files.
Its more like a temp directory now, but you should still be able to set the path of jniDir into your project where it can be commited.